### PR TITLE
feat(releasehealth): Add release health delegator [INGEST-253]

### PR DESCRIPTION
**Pull request by RaduW**
This PR adds a release health backend that delegates to the senssions and metrics backends and then compares the results.
It returns the value from the sessions backend.

[INGEST-498]

[INGEST-498]: https://getsentry.atlassian.net/browse/INGEST-498?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/sentry/release_health/duplex.py [hunk 1]
```diff
@@ -0,0 +1,763 @@
...
+from sentry.snuba.sessions_v2 import QueryDefinition
+
+DateLike = TypeVar("DateLike", datetime, str)
+
+
+class ComparatorType(Enum):
+    Counter = "counter"
+    Ratio = "ratio"
+    Quantile = "quantile"
+    Entity = "entity"
```
Comment (jjbayer) [inline src/sentry/release_health/duplex.py:43]: Would it make sense to rename `Entity` to `Exact`?

File: src/sentry/release_health/duplex.py [hunk 2]
```diff
@@ -0,0 +1,763 @@
...
+    def _dispatch_call(
+        self,
+        fn_name: str,
+        should_compare: Union[bool, Callable[[Any], bool]],
+        rollup: int,
+        schema: Optional[Schema],
+        *args,
+    ):
+        sessions_fn = getattr(self.sessions, fn_name)
+        ret_val = sessions_fn(*args)
```
Comment (jjbayer) [inline src/sentry/release_health/duplex.py:352]: Should we check a feature flag at this point, so we can turn on the double calling only for certain projects?

File: src/sentry/release_health/duplex.py (hunk 1 above)
Comment (RaduW) [reply src/sentry/release_health/duplex.py:43]: Yes, makes sense, let me think a bit on int

File: src/sentry/release_health/duplex.py (hunk 2 above)
Comment (RaduW) [reply src/sentry/release_health/duplex.py:352]: Yes.. not fully finished

**Review by untitaker — CHANGES_REQUESTED** [review]

File: src/sentry/release_health/duplex.py [hunk 3]
```diff
@@ -0,0 +1,912 @@
...
+        if should_compare:
+            copy = deepcopy(ret_val)
+            try:
+                metrics_fn = getattr(self.metrics, fn_name)
+                with timer("releasehealth.metrics.duration", tags=tags):
+                    metrics_val = metrics_fn(*args)
+            except Exception as ex:
+                self.log_exception(ex, fn_name, copy)
+            else:
+                errors = compare_results(copy, metrics_val, rollup, None, schema)
```
[the remark this answers is not in the document]
Comment (RaduW) [reply src/sentry/release_health/duplex.py:533]: ok, 
I could time `compare_results` but why ? 
Are you afraid it's going to take too much time while we test with both backends ?

File: src/sentry/release_health/duplex.py [hunk 4]
```diff
@@ -0,0 +1,912 @@
...
+            environments,
+        )
+
+    def check_has_health_data(
+        self, projects_list: Sequence[ProjectOrRelease]
+    ) -> Set[ProjectOrRelease]:
+        rollup = self.DEFAULT_ROLLUP  # not used
+        schema = {ComparatorType.Exact}
+        should_compare = datetime.utcnow() - timedelta(days=90) > self.metrics_start
+        organization = self._org_form_projects(projects_list)
```
[the remark this answers is not in the document]
Comment (RaduW) [reply src/sentry/release_health/duplex.py:702]: thanks, will fix all calls

File: src/sentry/release_health/duplex.py [hunk 5]
```diff
@@ -0,0 +1,912 @@
...
+
+    def __init__(self, schema, index_by=IndexBy):
+        self.child_schema = schema
+        if type(index_by) == str:
+            self.index_by: Optional[Callable[[Any], Any]] = lambda x: x.get(index_by)
+        else:
+            self.index_by = index_by
+
+
+Schema = Union[ComparatorType, List["Schema"], Mapping[str, "Schema"], Set["Schema"], ListSet]
```
[the remark this answers is not in the document]
Comment (RaduW) [reply src/sentry/release_health/duplex.py]: yep thanks, fixed

**Comment by untitaker** [comment]
we discussed in standup that this is ready for review

File: src/sentry/release_health/duplex.py [hunk 6]
```diff
@@ -0,0 +1,980 @@
...
+            "start": ComparatorType.DateTime,
+            "end": ComparatorType.DateTime,
+            "intervals": [ComparatorType.DateTime],
+            "groups": ListSet(
+                schema={
+                    "by": ComparatorType.Ignore,
+                    "series": schema_for_series,
+                    "totals": schema_for_totals,
+                },
+                index_by="by",
```
Comment (jjbayer) [inline src/sentry/release_health/duplex.py:637]: Does this actually work? `groups[0]["by"]` is itself a dict, so you won't be able to use it as a dict key here:
https://github.com/getsentry/sentry/pull/29731/files#diff-90f1dc7390bf824e8bbaf2c0532cf4416195cf9134733a24064b2ec152cc4e87R380

Might need to set `index_by=lambda x: tuple(sorted(x["by"].items()))` like here:

https://github.com/getsentry/sentry/blob/9a0c39c9870cbd426e250dc5ffab7875bb601e72/tests/snuba/api/endpoints/test_organization_sessions.py#L19-L20

Comment (RaduW) [reply src/sentry/release_health/duplex.py:636]: thanks, didn't relaise that "by" is a dict (thougnt is a string)

File: src/sentry/release_health/duplex.py [hunk 7]
```diff
@@ -0,0 +1,983 @@
...
+            self.index_by: Callable[[Any], Any] = lambda x: x.get(index_by)
+        else:
+
+            self.index_by = cast(Callable[[Any], Any], index_by)
+
+
+Schema = Union[ComparatorType, List[Any], Mapping[str, Any], Set[Any], ListSet]
+
+
+def _get_calling_method() -> str:
```
[the remark this answers is not in the document]
Comment (RaduW) [reply src/sentry/release_health/duplex.py:83]: not anymore, now passing the method name explicitly... will remove, thanks

File: src/sentry/release_health/duplex.py [hunk 8]
```diff
@@ -0,0 +1,912 @@
...
+    if type(sessions) == str:
+        try:
+            sessions_d = parser.parse(sessions)
+            metrics_d = parser.parse(metrics)
+            dd = abs(sessions_d - metrics_d)
+        except parser.ParserError:
+            return f"field {path} could not parse dates sessions={sessions}, metrics={metrics}"
+    else:
+        dd = abs(sessions - metrics)
+    if dd > timedelta(seconds=rollup):
```
[the remark this answers is not in the document]
Comment (RaduW) [reply src/sentry/release_health/duplex.py:144]: yes... will see if it needs a different adjustment.

File: src/sentry/release_health/duplex.py [hunk 9]
```diff
@@ -0,0 +1,983 @@
...
+    ) -> ReleaseHealthResult:
+        if rollup is None:
+            rollup = 0  # force exact date comparison if not specified
+
+        sessions_fn = getattr(self.sessions, fn_name)
+        tags = {"method": fn_name}
+        with timer("releasehealth.sessions.duration", tags=tags):
+            ret_val = sessions_fn(*args)
+
+        if organization is None or not features.has("organizations:release-health-check-metrics"):
```
[the remark this answers is not in the document]
Comment (RaduW) [reply src/sentry/release_health/duplex.py:508]: It's a bug ... I forgot to pass the org, thanks.

File: src/sentry/release_health/duplex.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (RaduW) [reply src/sentry/release_health/duplex.py]: done

**Review by jjbayer — APPROVED** [review]

**Review by untitaker — APPROVED** [review]

File: src/sentry/release_health/duplex.py [hunk 10]
```diff
@@ -633,7 +629,7 @@ def run_sessions_query(
                     "series": schema_for_series,
                     "totals": schema_for_totals,
                 },
-                index_by="by",
+                index_by=dict_to_tuple,
```
[the remark this answers is not in the document]
Comment (RaduW) [reply src/sentry/release_health/duplex.py]: yes, fixed the index_by function and added UT for this case, thanks

**Pull request merged**